libbeat/publisher/pipeline/output_test.go

```diff
@@ -0,0 +1,260 @@
+			seedPRNG(t)
+
+			err := quick.Check(func(i uint) bool {
+				numBatches := 300 + (i % 100) // between 300 and 399
+
+				wqu := makeWorkQueue()
+				client := ctor(0)
+				makeClientWorker(nilObserver, wqu, client)
+
+				numEvents := atomic.MakeUint(0)
+				for batchIdx := uint(0); batchIdx <= numBatches; batchIdx++ {
+					batch := randomBatch(50, 150, wqu)
+					numEvents.Add(uint(len(batch.Events())))
+					wqu <- batch
+				}
+
+				// Give some time for events to be published
+				timeout := 20 * time.Second
+
+				// Make sure that all events have eventually been published
+				return waitUntilTrue(timeout, func() bool {
+					return numEvents.Load() == client.Published()
+				})
+			}, nil)
+
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestPublishWithClose(t *testing.T) {
+	tests := map[string]func(uint) publishCountable{
+		"client":         newMockClient,
+		"network_client": newMockNetworkClient,
+	}
+
+	const minEventsInBatch = 50
+
+	for name, ctor := range tests {
+		t.Run(name, func(t *testing.T) {
+			seedPRNG(t)
+
+			err := quick.Check(func(i uint) bool {
+				numBatches := 1000 + (i % 100) // between 1000 and 1099
+
+				wqu := makeWorkQueue()
+				numEvents := atomic.MakeUint(0)
+
+				var wg sync.WaitGroup
+				for batchIdx := uint(0); batchIdx <= numBatches; batchIdx++ {
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+
+						batch := randomBatch(minEventsInBatch, 150, wqu)
+						numEvents.Add(uint(len(batch.Events())))
+						wqu <- batch
+					}()
+				}
+
+				// Publish at least 1 batch worth of events but no more than 20% events
+				publishLimit := uint(math.Min(minEventsInBatch, float64(numEvents.Load())*0.2))
+				client := ctor(publishLimit)
+				worker := makeClientWorker(nilObserver, wqu, client)
+
+				// Allow the worker to make *some* progress before we close it
+				timeout := 10 * time.Second
+				progress := waitUntilTrue(timeout, func() bool {
+					return client.Published() >= publishLimit
+				})
+				if !progress {
+					return false
+				}
+
+				// Close worker before all batches have had time to be published
+				err := worker.Close()
+				require.NoError(t, err)
+
+				published := client.Published()
+
+				// Start new worker to drain work queue
+				client = ctor(0)
+				makeClientWorker(nilObserver, wqu, client)
+				wg.Wait()
+
+				// Make sure that all events have eventually been published
+				timeout = 20 * time.Second
+				return waitUntilTrue(timeout, func() bool {
+					total := published + client.Published()
+					return numEvents.Load() == total
+				})
+			}, &quick.Config{MaxCount: 50})
+
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+type publishCountable interface {
+	outputs.Client
+	Published() uint
+}
+
+func newMockClient(publishLimit uint) publishCountable {
+	return &mockClient{publishLimit: publishLimit}
+}
+
+type mockClient struct {
+	mu           sync.RWMutex
+	publishLimit uint
+	published    uint
```

Comment:
maybe we can excert more control by replacing `published` and `publishLimit` with a `func`. The `func` could decide when to block and when to continue. This way the test logic would be more contained within the test. WDYT?

Reply:
Implemented mockable publish behavior function in 508b60687.